rts/Sim/Units/Scripts/CobInstance.cpp

```diff
@@ -701,13 +701,19 @@ void CCobInstance::ThreadCallback(ThreadCallbackType type, int retCode, int cbPa
 
 void CCobInstance::Signal(int signal)
 {
-	for (int threadID: threadIDs) {
-		CCobThread* t = cobEngine->GetThread(threadID);
-
-		if ((signal & t->GetSignalMask()) == 0)
-			continue;
-
-		t->SetState(CCobThread::Dead);
+	for (size_t i = 0; i < threadIDs.size();) {
+		CCobThread* t = cobEngine->GetThread(threadIDs[i]);
+		if ((signal & t->GetSignalMask()) == 0) {
+			// not matching signal, let's go to next thread on the list
+			++i;
+		} else {
+			t->SetState(CCobThread::Dead);
+			// when we drop thread, we don't go to next element, as thread
+			// destructor removes it from the vector we iterate over.
```

Comment:
Since we're relying on vector internals, could there be some sort of basic check?
```
static_assert(std::is_same<decltype(threadIDs), std::vector<int>>::value,
    "container type changed, make sure this loop is still correct");
```

Reply:
We are depending on stable positions, which isn't vector specific, it's more about how are we doing the removal, not really that much type dependent, so I don't think it's that useful.

Instead I've added comment in the RemoveThreadID about what guarantee it needs to hold.